search: fork/archived UI notifications

rvantonder opened the pull request:
This adds a notification as seen below. This shows up if a fork (resp. archive) is excluded  _only if_ the `fork:` (resp. `archive:`) option isn't explicitly specified in the query.

![Screen Shot 2020-05-12 at 11 01 39 PM](https://user-images.githubusercontent.com/888624/81776894-8ed39600-94a4-11ea-921f-73e8b919a411.png)

I plan to still add an e2e web regression test and some backend test, though I would much prefer an integration test for it.

Notes:

- On hover, it says to `add fork:yes` followed by the list of repos. There isn't a lot of freedom to format this, and it will overflow on the width of the hover box if the message is longer. I think the current message is OK.

- A repo may be a fork and an archive, or only a fork/only an archive. It's best if we treat the labels separately as far as showing the notification

**Backend implementation stuff**

- At first I thought to do a query of all repos including forks/archives, and then filter out the forks/archives. But, I discovered that the standard DB call will not populate those fields/labels for repos due to perf issues. Instead of iterating over the entire list and then 'hydrating' whether they are forks or repos, I instead take the original query and modify it to make two additional DB calls where I set `fork:only` and then `archived:only`, and put those matching repos in the 'excluded' list. My intuition is that this is faster than calling `Get(...)` for each repo and then checking fork/archived. The additional DB calls run in the typical case when `fork`/`archived` unspecified, but not otherwise.

- I had to refactor tests that relied on a mocked DB `List` call. These tests want certain invariants to hold for the DB options when matching repos. The current/previous 'easy' way to do this is just to accept all the DB options as default and set the ones that matter. When we changed the default of fork/archive matching in a previous PRs (e.g., https://github.com/sourcegraph/sourcegraph/pull/8739), I had to change these tests to now use the new `NoFork`... default option. Now, these values can change since there are multiple calls to DB `List`. The answer is that these tests should not be relying on such strong invariants on the fork/archive inclusion/exclusion logic because that's not what they're testing. I have updated the tests to check for weaker invariants instead.

- This will conflict with refactor #10489. I'd appreciate if we could hold off on merging #10489 until after branch cut, since I'd like to get this feature in for 3.16 without issues that might crop up due to refactor and/or adjusting to it.

rvantonder commented:
This works on manual testing as expected, but I'm getting `TypeError: Cannot read property 'filter' of undefined` for the `props.results.excluded.filter(repo => repo.isFork)` call in the tests. Not sure what I'm doing wrong (why does `.map(...)` work elsewhere?), but it's probably something silly.

keegancsmith commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -1362,6 +1362,11 @@ type SearchResults {
+    #
+    # This list currently applies only to fork and archived repositories, which are excluded by
+    # default in search queries.
+    excluded: [Repository!]!
```
this list could be quite large? Do we really need the whole list to add the label? Surely we can do it with a db.list with a `select name, fork, count(1) from repos where $SEARCH_CONDS and fork group by fork`. This suggestion is me thinking from a perf perspective.

slimsag replied:
This is a good point; just including the count would be sufficient.

rvantonder replied:
Is the proposal that this GQL type reflects only the count? How would the count of forks versus archived repos when only exposing count? I.e., `excludedForks: int` and `excludedArchived: int`? Or is the proposal here to have a type be something that is `[names], count` or something else?

slimsag replied:
>  I.e., excludedForks: int and excludedArchived: int

Yes, this is what I was thinking.

keegancsmith commented on cmd/frontend/graphqlbackend/search.go:
```diff
@@ -365,9 +365,37 @@ func exactlyOneRepo(repoFilters []string) bool {
+func (r *searchResolver) computeExcludedRepositories(ctx context.Context, op resolveRepoOp) (excludedRepoRevs []*search.RepositoryRevisions) {
+	forkStr, _ := r.query.StringValue(query.FieldFork)
+	fork := parseYesNoOnly(forkStr)
+	if fork == Invalid {
```
if the repo query is exact we search forks. Have you tested searching a fork? Interested to know if we will search it but then say it has been excluded.

rvantonder replied:
Correct!! Needs to be part of the test.

keegancsmith commented:
Didn't review web code, so I guess get approval for that. They could probably also help you with your testing woes.

unknwon commented on cmd/frontend/graphqlbackend/search_results_test.go:
```diff
@@ -30,6 +30,12 @@ import (
 	"github.com/sourcegraph/sourcegraph/schema"
 )
 
+func assert_equal(t *testing.T, got, want interface{}) {
```
```suggestion
func assertEqual(t *testing.T, got, want interface{}) {
	t.Helper()
```

:)

rvantonder replied:
D:

unknwon commented on cmd/frontend/graphqlbackend/search_results.go:
```diff
@@ -1678,7 +1688,14 @@ func (r *searchResolver) doResults(ctx context.Context, forceOnlyResultType stri
+		len(common.missing),
+		len(common.excluded),
+		len(common.timedout))
+	log15.Info("0.", "si", len(common.excluded))
```
Debugging leftover?

rvantonder replied:
D:

unknwon commented on cmd/frontend/graphqlbackend/search_results_test.go:
```diff
@@ -30,6 +30,12 @@ import (
 )
 
+func assert_equal(t *testing.T, got, want interface{}) {
+	if !reflect.DeepEqual(got, want) {
+		t.Fatalf("got %+v, want %+v", got, want)
+	}
```
```suggestion
	if diff := cmp.Diff(want, got); diff != "" {
		t.Fatalf("(-want +got):\n%s", diff)
	}
```

uwedeportivo commented:
Dear all,

This is your release captain speaking. 🚂🚂🚂

Branch cut for the **3.16 release is scheduled for tomorrow**.

Is this issue / PR going to make it in time? Please change the milestone accordingly.
When in doubt, reach out!

Thank you

rvantonder commented:
>  At a high level I would approached this by adding a new dynamic filter. ... could be done in a way which is independent of "resolveRepositories". Did you consider this approach?

This PR is based off of the suggestion in https://github.com/sourcegraph/sourcegraph/issues/9928#issuecomment-624997214

I don't think the dynamic filter solution addresses all the concerns raised in https://github.com/sourcegraph/sourcegraph/issues/9928, namely:
- we should be clear about e.g. how many forks were excluded
- the filter would not suffice in the context of https://github.com/sourcegraph/sourcegraph/issues/10313#issuecomment-625451535, since this is only a  "semblance of 'there are possibly more results omitted here'" and currently used as a stop gap.

Basically, I dismissed this approach because I'm convinced based on the discussion it doesn't address the issue. I think it's useful to know that forks/archives were omitted instead of preemptively adding dynamic filters, and as such I don't see any way but to rely on some sort of `resolveRepositories`-like call to know that (unless I'm missing something about what is implied by the approach you had in mind).

> this list could be quite large? Do we really need the whole list to add the label? Surely we can do it with a db.list with a `select name, fork, count(1) from repos where $SEARCH_CONDS and fork group by fork`. This suggestion is me thinking from a perf perspective.

In the context of the parent issue, the concern is to surface more complete knowledge of what was omitted. This could be "at least N fork repos were excluded" for N = 1, but I don't know that including partial information of excluded repos is a good perf <-> UX tradeoff. @slimsag thoughts?

rvantonder commented on web/src/search/results/SearchResultsInfoBar.tsx:
```diff
@@ -118,6 +120,32 @@ export const SearchResultsInfoBar: React.FunctionComponent<SearchResultsInfoBarP
+                                    {pluralize('repository', repos.length, 'repositories')} excluded
+                                </span>
+                            </div>
+                        ))(props.results.excluded.filter(repo => repo.isArchived))}
```
I wanted to define variables but I couldn't find a place to do it in this function, since it's just expressions. Where would be a suitable place to do this?

felixfbecker replied:
In VS Code, just put the cursor into the `=>` arrow in L80, hit cmd+. and select "add braces to arrow function". Then you can declare variables in the body.

More info on the syntax: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#Shorter_functions

rvantonder replied:
Ah ha!

keegancsmith commented:
Well shit I just wrote a long response and it got lost somehow. So just gonna write a very brief response:

The approach I suggested is exact and fast (it includes a count). Dynamic filters can do DB queries/etc. The UI you have is much better though, but maybe the ideas can be extended.

In the past our perf work for large amounts of repos was all about the DB. Listing repos is slow, that is why we have the funky hydration logic. Your query asks for forked/archived which can make it very slow. Given it is the day of branch cut can you add a feature flag (default on) for populating this list. I am concerned about performance.

Can you mark the field you add to graphql as experimental, so we have freedom to change it next release if we decide to.

GraphQL allows querying the size of a list and just fetching N items. This may be useful, but I'm unsure if our graphql server library makes it possible to do efficiently.

rvantonder commented:
>  got lost somehow. 

:'( 

> The approach I suggested is exact and fast (it includes a count). Dynamic filters can do DB queries/etc

I'm not familiar with the part of the code that will give me this count without querying here. Whenever we have a chance, feel free to point that out. Updating the UI code shouldn't be a big deal if I can extract this value on the frontend.

> I am concerned about performance.

I mean, yes, me too. But I don't see another way to do this. Also, previously, before excluding forks by default, the DB List would run including forks/archives, so we would conceptually end up with the same size of the list and similar DB queries that check `Fork/`Archive` fields in the DB. I realize multiple queries are not the same, and I don't know if that intuition checks out. Other than that and the slow DB lookup concern, we don't need to send the result of DB List over GQL in the `exclude` member, and just clamp it.

> Can you mark the field you add to graphql as experimental, so we have freedom to change it next release if we decide to.

> GraphQL allows querying the size of a list and just fetching N items. This may be useful, but I'm unsure if our graphql server library makes it possible to do efficiently.

I'll see what I can do tomorrow before branch cut, but I'm not going to go to nuts here, I've already overcommitted to get to this point. I'd rather just not merge rushing for branch cut and continue in the next iteration.

keegancsmith commented:
> I'll see what I can do tomorrow before branch cut, but I'm not going
> to go to nuts here, I've already overcommitted to get to this
> point. I'd rather just not merge rushing for branch cut and continue
> in the next iteration.

Agreed, don't rush.

slimsag commented:
Yeah Keegan makes some good points here, I apologize for not thinking of that ahead of time.

No need to land this in 3.16.0, we can do a patch release once this is solid and ready. It's worth taking the time to do it right (but worth keeping as a high priority, too, of course).

rvantonder commented:
Based on the current discussion, I'm fuzzy on how to proceed. It's sounds like there are other ways to achieve what we want. Directions:

1. Change the schema to `excludedForks: int`, `excludedArchived: int`, no repo names. 
Add a backend DB call that only gets the count for these respectively. Where/how exactly we do this I'm not exactly sure yet, so I suppose the proposal is to add a helper function that does this? I'm inclined to add this in the vicinity of `resolveRepos` (see below). This direction sounds like we don't need to expose a parameter N then.

1. Use current state of PR, but add ability to fetch only N items from `excluded` as a GQL parameter. Mark the `excluded` field as experimental. Accept a parameter N in the GQL `excluded` endpoint for the size of the list. 

1. Alternatively, via dynamic filter logic which queries DB based on suggestion:
> exact and fast (it includes a count). Dynamic filters can do DB queries/etc. 

I looked briefly at the [dynamic filter code](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@48cb2bc0136f768ff18577ab4c6a6e11b431cdb8/-/blob/cmd/frontend/graphqlbackend/search_results.go#L221), it doesn't seem to be doing any DB-related calls currently so that would need to be added? I think the `DynamicFilter` endpoint is too closely attached to the UI chip representation and seems orthogonal to addressing the perf issue. I think the UI in the status bar of the original PR is better suited than the dynamic filter/chip UI. Thus, whatever count/DB lookup we would have done to add for a dynamic filter can be done in the part that looks up forks/archives around the area of resolveRepos. I sympathize with not wanting to bloat resolveRepos due to the complexity, but doing the lookup in that vicinity feels like a better, more UI-independent place for this to live. I can try separate this logic more to help with that, but will likely take more time.

Maybe we can align on option (1) here @slimsag and @keegancsmith?

slimsag commented:
I think (3) will be harder to implement and is further away from your current PR, so I would advise not going with that option for now at least.

I think (2) would be OK, and as Keegan noted in his inline comment it is possible for us to modify the `resolveRepositories` db query to only select N "excluded" repositories from the DB, but I think this will be tricky/difficult to implement unless you hard-code the number to e.g. 5 and then the UI would either need to display "5+ forked repositories excluded" or separately query the total number. My only argument against this solution is that I think it will involve making more changes and be more involved than it sounds at first glance.

I think (1) is a good solution and should be easiest to implement, you would just need to modify `resolveRevision` to [call `db.Repos.Count`](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24%408b86a3c+db.Repos.Count&patternType=regexp) instead of `db.Repos.List` which should make it much more efficient to calculate.

In any case, I think me and @keegancsmith both ultimately trust your judgement on the best way to proceed here :)

keegancsmith commented:
My personal preference is to avoid expanding our graphQL API, especially if it involves harding things like "archived" and "forked". But to be honest it is the most practical way to achieve what we want, so lets do that.

I think the problem with dynamic filters is the UI isn't as nice, so it makes sense to make it first class. So go for (1)

slimsag commented:
Give me a ping when this is ready to be reviewed again

rvantonder commented:
OK finally converged on this. Solution:

- Only do accounting for the number of repos excluded
- Use `Repos.Count` in `computeExcludedRepositories` (all affected tests  mock this function now)
- Expose the counts through `DynamicFilter` as proposed. In the webapp, pull these values out of the `DynamicFilter` to show the same UI notification (and _also_ the dynamic filter chips, order is nondeterministic)
- No changes to GQL API (it's fairly straightforward to get the info out of dynamic filter values)

![Screen Shot 2020-05-18 at 4 22 12 PM](https://user-images.githubusercontent.com/888624/82268417-c8345780-9923-11ea-8fbb-f341bb06186a.png)

This improves on solution (1) above because there are no changes to the GQL API now by using values exposed via DynamicFilter. PR Needs to be reviewed again because of the above substantial changes. Webapp (cc @lguychard) is only minor changes. PTAL.

slimsag commented:
> Expose the counts through DynamicFilter as proposed. In the webapp, pull these values out of the DynamicFilter to show the same UI notification (and also the dynamic filter chips, order is nondeterministic)

Having both of these is awesome!

slimsag commented on cmd/frontend/graphqlbackend/search.go:
```diff
@@ -365,9 +366,45 @@ func exactlyOneRepo(repoFilters []string) bool {
+		selectForks := op
+		selectForks.OnlyForks = true
+		selectForks.NoForks = false
+		numExcludedForks, _ = db.Repos.Count(ctx, selectForks)
```
Although unlikely, I could see this failing in some cases (e.g. if we bork the SQL query) and it going unnoticed and being painful to debug. Maybe check the error and `log15` it instead?

rvantonder replied:
ack

slimsag commented on CHANGELOG.md:
```diff
@@ -13,6 +13,8 @@ All notable changes to Sourcegraph are documented in this file.
 
 ### Added
 
+- The search results page now shows a small UI notification if either repository forks or archives are excluded, when `fork` or `archived` options are not explicitly set. [#10624](https://github.com/sourcegraph/sourcegraph/pull/10624)
```
Q: If forks are excluded and `fork:no` is explicitly set, do the search filter chip `fork:yes` show? (ideally I think it would, not important just curious)

rvantonder replied:
Any time `fork` or `archived` is explicitly set, there is no check on how many forks or archives we potentially miss (that's what the unit tests check). This implies that no UI notifications/chips are shown.

To show `fork:yes`, we would have to inspect the value of the query and run the part in `computeExcludedRepos` based on that (slightly more additional logic).

slimsag replied:
Got it, makes sense. No need to do that I think, was just curious.

slimsag commented on cmd/frontend/graphqlbackend/search.go:
```diff
@@ -365,9 +366,45 @@ func exactlyOneRepo(repoFilters []string) bool {
+		selectArchived := op
+		selectArchived.OnlyArchived = true
+		selectArchived.NoArchived = false
+		numExcludedArchived, _ = db.Repos.Count(ctx, selectArchived)
```
with this error, too

rvantonder replied:
ack

slimsag commented on cmd/frontend/graphqlbackend/search.go:
```diff
@@ -365,9 +366,45 @@ func exactlyOneRepo(repoFilters []string) bool {
 }
 
+// A type that counts how many repos with a certain label were excluded from search results.
+type excludedRepos = struct {
```
Declaring a struct type with `=` has weird subtle differences you probably do not want. I didn't realize this was possible until I saw it in your PR just now.

It declares the type equal to an anonymous struct: https://play.golang.org/p/tJo403zJIdG

```suggestion
type excludedRepos struct {
```

rvantonder replied:
this is probably a typo :⭕

slimsag replied:
Oh, ahahaha, It's actually a [type alias](https://golang.org/doc/go1.9#language) [declaration](https://golang.org/ref/spec#Type_declarations) to an anonymous struct which I guess in the worst case could end up acting like a pseudo-generic type, interesting: https://play.golang.org/p/tT7FoHRsqMN

rvantonder replied:
Oh I'm familiar with the type alias syntax but haven't used it to directly alias based of a struct, that seems weird.

slimsag replied:
Yeah it'd definitely be an abuse of the type-system to use it in that way ;) I just though it was funny

rvantonder replied:
[Some protobuf stuff I guess ¯\\\_(ツ)_/¯](https://sourcegraph.com/search?q=%2Ftype+%5Cw%2B+%3D+struct+%7B%2F+lang%3Ago+count%3A1000&patternType=regexp#36)

slimsag replied:
Wow, of course it would be used there.. I guess someone _did_ have a real use for it
